### PR TITLE
perf(math): optimize usage of is one

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR optimizes `IsOne()` by avoiding montgomery reduction and optimizes some functions by not multiplying the operand if it is one.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

File: tachyon/math/base/semigroups.h [hunk 1]
```diff
@@ -167,18 +167,17 @@ class MultiplicativeSemigroup {
...
-        base::GetNumElementsPerThread(ret, kDefaultParallelThreshold);
-    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
-      MulResult pow = generator.Pow(i);
-      if (!c.IsOne()) pow *= c;
-      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
-        ret[j] = pow;
-        pow *= generator;
-      }
-    }
+    std::vector<MulResult> ret = base::CreateVector(size, MulResult::Zero());
```
Comment (TomTaehoonKim) [inline tachyon/math/base/semigroups.h:170]: Can you make this change a separate commit?

File: tachyon/math/base/semigroups.h [hunk 2]
```diff
@@ -313,17 +312,16 @@ class AdditiveSemigroup {
...
-    size_t num_elems_per_thread =
-        base::GetNumElementsPerThread(ret, kDefaultParallelThreshold);
-    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
-      AddResult scalar_mul = generator.ScalarMul(i);
-      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
-        ret[j] = scalar_mul;
-        scalar_mul += generator;
-      }
-    }
+    std::vector<AddResult> ret = base::CreateVector(size, AddResult::Zero());
```
Comment (TomTaehoonKim) [inline tachyon/math/base/semigroups.h:315]: Can you make this change a separate commit?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

File: tachyon/math/base/semigroups.h (hunk 2 above)
Comment (chokobole) [reply tachyon/math/base/semigroups.h]: Since we are talking about removing `base::CreateVector()` which is similar to the constructor of `std::vector<>`, I just change it to `std::vector<AddResult> ret(size);`

File: tachyon/math/base/semigroups.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/base/semigroups.h]: Since we are talking about removing `base::CreateVector()` which is similar to the constructor of `std::vector<>`, I just change it to `std::vector<MulResult> ret(size);`

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM
